src/audio/dai-zephyr.c

```diff
@@ -896,21 +914,14 @@ int dai_zephyr_params(struct dai_data *dd, struct comp_dev *dev,
 		return -EINVAL;
 	}
 
-	if (dev->direction == SOF_IPC_STREAM_PLAYBACK)
-		dd->local_buffer = list_first_item(&dev->bsource_list,
-						   struct comp_buffer,
-						   sink_list);
-	else
-		dd->local_buffer = list_first_item(&dev->bsink_list,
-						   struct comp_buffer,
-						   source_list);
-
 	/* check if already configured */
 	if (dev->state == COMP_STATE_PREPARE) {
 		comp_info(dev, "dai_zephyr_params() component has been already configured.");
 		return 0;
 	}
 
```

Comment:
@kv2019i this is an excellent find. In the case of the DaI  copier with multiple sinks, we connect all the sinks during hw_params and then during trigger we trigger all the pipelines. And depending on which order the sinks are connected, it could either be the smart_amp or the gain module that goes first. But I dont think the right solution is to set the local_buffer during copy every time. Instead we need to identify the buffer in the capture path to the gain and set that as the local_buffer during params. This is needed because we use the local_buffer frame_fmt to choose the processing function for the DAI copy.

Reply:
@kv2019i I thought about this a bit more and I think a more reasonable solution would be to remove this altogether https://github.com/thesofproject/sof/blob/38f3f5d4119e382c8cee99bbc6ec1d3fff407f93/src/audio/dai-zephyr.c#L267.

And let the loop below handle all sinks. This way all the sinks that are not connected won't be used also.

Reply:
should also need remove line L269? totally use loop?

Reply:
@ranj063 @btian1 But L267 is outside IPC4 ifdef, so this would seem to break IPC, right? 

I don't think setting local_buf at every copy is very expensive as we are taking the first item on the list, so this should be easy case for the compiler to handle. But I do agree with @lyakh this feels brittle. We should have some explicit action when modules are disconnected, both for functional correctness and to make the code easier to follow. Now storing a local copy of bsource_list/bsink_list sounds like a very, very, very bad idea when ipc_comp_disconnect() may be called at any time.

But should we have a simple fix first, or should I send a version that adds bind/unbind to copier right away?


@lyakh yes, this feels

Reply:
@ranj063 Hmm, I tried a simple variant where I pick the local_buffer from the sink that is on the same pipeline (as components of the same pipeline are guaranteed to be persistent while pipeline is playing). But I seem to be hitting errors that need more debug. Plus, not sure whether we can assume copier at least one sink connected to a component in the same pipeline. Probably always the case in practise, but this configuration is not really forbidden either.

Reply:
Addressed dd->process reconfig in V2 and implemented unbind. I'll close this conversation, please comment on the new code.